### PR TITLE
Amount as Number (SIRI-719)

### PR DESCRIPTION
**Pull request by jakobvogel**
> https://scireum.myjetbrains.com/youtrack/issue/SIRI-719

`Amount` becomes a `Number`, yay!

### CODE REVIEW

File: src/main/java/sirius/kernel/commons/Value.java [hunk 1]
```diff
@@ -893,11 +887,8 @@ public Integer getInteger() {
             if (data instanceof Integer integer) {
```
Comment (jakobvogel) [inline src/main/java/sirius/kernel/commons/Value.java:887]: Left this as this is exactly the return type. Ok?

File: src/main/java/sirius/kernel/commons/Value.java [hunk 2]
```diff
@@ -972,8 +954,8 @@ public Long getLong() {
             if (data instanceof Long longValue) {
```
Comment (jakobvogel) [inline src/main/java/sirius/kernel/commons/Value.java:954]: Left this as this is exactly the return type. Ok?

File: src/test/java/sirius/kernel/commons/ValueSpec.groovy [hunk 3]
```diff
@@ -168,6 +168,77 @@ class ValueSpec extends BaseSpecification {
...
+        99999.0           | new Long(99999l)
+        new Double(99999) | new Long(99999l)
+        Amount.of(99999)  | new Long(99999l)
+        "99999"           | new Long(99999l)
+        "Keine Zahl"      | null
+    }
+
+    def "Casting to doubles works"() {
+        expect:
+        Value.of(input).asDouble(1234.56) == output
```
Comment (jakobvogel) [inline src/test/java/sirius/kernel/commons/ValueSpec.groovy:229]: Is this a valid double comparison in Groovy? There should be an ε

Comment (andyHa) [reply src/test/java/sirius/kernel/commons/ValueSpec.groovy:229]: Doubles.EPSILON?

**Review by andyHa — APPROVED** [review]

File: src/main/java/sirius/kernel/commons/Amount.java [hunk 4]
```diff
@@ -244,6 +244,26 @@ public BigDecimal getAmount() {
         return value;
     }
 
+    @Override
+    public int intValue() {
+        return value.intValue();
```
Comment (mkeckmkeck) [inline src/main/java/sirius/kernel/commons/Amount.java:249]: Dont't we require null checks + fallback here e.g. for Amount.NOTHING?

**Review by idlira — APPROVED** [review]

File: src/main/java/sirius/kernel/commons/Amount.java (hunk 4 above)
Comment (jakobvogel) [reply src/main/java/sirius/kernel/commons/Amount.java:249]: Hmm, to me it feels ok to have an NPE here when treating an `Amount.NOTHING` as `Number`, but I am open for discussion. For now, I have added such cases to the test, and I have reinforced the casting in `Value`

File: src/main/java/sirius/kernel/commons/Value.java (hunk 1 above)
Comment (sabieber) [reply src/main/java/sirius/kernel/commons/Value.java:887]: For simplicity sake I would include it in the Number case and remove this if. When looking at the Implementation of the Integer or Long class they do exactly the same

File: src/main/java/sirius/kernel/commons/Value.java (hunk 2 above)
Comment (sabieber) [reply src/main/java/sirius/kernel/commons/Value.java:954]: For simplicity sake I would include it in the Number case and remove this if. When looking at the Implementation of the Integer or Long class they do exactly the same

**Review by sabieber — APPROVED** [review]

File: src/main/java/sirius/kernel/commons/Amount.java (hunk 4 above)
Comment (andyHa) [reply src/main/java/sirius/kernel/commons/Amount.java:249]: Return Integer.MIN_VALUE - this should probably the most unexpected behaviour :-P

Comment (andyHa) [reply src/main/java/sirius/kernel/commons/Amount.java:249]: One could return an IllegalStateException (?) as this would feel less "random" ;-)

Comment (sabieber) [reply src/main/java/sirius/kernel/commons/Amount.java:249]: Right, that's actually a very valid point.

I quite like the idea of an exception, additionally we maybe could expand the filled check methods (if they currently don't have a code path for Amounts) to act like the Value is empty when the Amount is empty

File: src/main/java/sirius/kernel/commons/Amount.java [hunk 5]
```diff
@@ -246,21 +246,37 @@ public BigDecimal getAmount() {
...
+            throw new IllegalStateException("The value is empty.");
+        }
+
         return value.intValue();
     }
 
     @Override
     public long longValue() {
+        if (value == null) {
+            throw new IllegalStateException("The value is empty.");
```
Comment (idlira) [inline src/main/java/sirius/kernel/commons/Amount.java:259]: Extract in createIllegalStateException? (or define a constant for the string)?

**Pull request merged**